main.go

```diff
@@ -118,6 +118,10 @@ func main() {
 	opts.Programs.KernelTypes = btfSpec
 	opts.Programs.LogLevel = ebpf.LogLevelInstruction
 	opts.Programs.LogSize = ebpf.DefaultVerifierLogSize * 100
+	opts.Maps.PinPath = "/sys/fs/bpf/pwru"
+	if err = os.MkdirAll(opts.Maps.PinPath, 0755); err != nil {
```

Comment:
The pinned maps should be cleared when exit by `os.RemoveAll(opts.Maps.PinPath)`.

However, what if `/sys/fs/bpf` is not mounted?

Would it be better to add these maps to `replacedMaps` at https://github.com/cilium/pwru/blob/dbaa40676fd3d41bdb9c387f9251119c8649d785/internal/pwru/tracing.go#L134

Reply:
replacedMap doesn't work, this is my patch
```diff
diff --git a/bpf/kprobe_pwru.c b/bpf/kprobe_pwru.c
index 8e66152..9b3eb8c 100644
--- a/bpf/kprobe_pwru.c
+++ b/bpf/kprobe_pwru.c
@@ -488,6 +488,8 @@ cont:
 	event->ts = bpf_ktime_get_ns();
 	event->cpu_id = bpf_get_smp_processor_id();
 
+	event->print_stack_id = tracked_by;
+
 	return true;
 }
 
diff --git a/internal/pwru/output.go b/internal/pwru/output.go
index 254b906..d3b128a 100644
--- a/internal/pwru/output.go
+++ b/internal/pwru/output.go
@@ -435,6 +435,7 @@ func (o *output) Print(event *Event) {
 		fmt.Fprintf(o.writer, "%s", getShinfoData(event, o))
 	}
 
+	fmt.Fprintf(o.writer, " %d", event.PrintStackId)
 	fmt.Fprintln(o.writer)
 }
 
diff --git a/internal/pwru/tracing.go b/internal/pwru/tracing.go
index f1751f1..b4e550c 100644
--- a/internal/pwru/tracing.go
+++ b/internal/pwru/tracing.go
@@ -132,8 +132,13 @@ func (t *tracing) trace(coll *ebpf.Collection, spec *ebpf.CollectionSpec,
 	// Reusing maps from previous collection is to handle the events together
 	// with the kprobes.
 	replacedMaps := map[string]*ebpf.Map{
-		"events":          coll.Maps["events"],
-		"print_stack_map": coll.Maps["print_stack_map"],
+		"events":            coll.Maps["events"],
+		"skb_addresses":     coll.Maps["skb_addresses"],
+		"veth_skbs":         coll.Maps["veth_skbs"],
+		"skb_stackid":       coll.Maps["skb_stackid"],
+		"stackid_skb":       coll.Maps["stackid_skb"],
+		"xdp_dhs_skb_heads": coll.Maps["xdp_dhs_skb_heads"],
+		"print_stack_map":   coll.Maps["print_stack_map"],
 	}
 	if outputSkb {
 		replacedMaps["print_skb_map"] = coll.Maps["print_skb_map"]

```

Reply:
We should confirm it by `bpftool prog`:

```bash
# bpftool p
3958: kprobe  name kprobe_multi_skb_3  tag 01ea9039709969c3  gpl
        loaded_at 2024-11-27T10:04:12+0800  uid 0
        xlated 7528B  jited 4003B  memlock 12288B  map_ids 24586,24592,24593,24589,24583,24587,24585,24594,24595,24588,24584,24590
        btf_id 27403
3962: tracing  name fentry_xdp  tag 5dae1059a9cdb3e1  gpl
        loaded_at 2024-11-27T10:04:12+0800  uid 0
        xlated 4544B  jited 2524B  memlock 8192B  map_ids 24602,24603,24584,24592
        btf_id 27417
```

Then, with this diff
```diff
diff --git a/internal/pwru/tracing.go b/internal/pwru/tracing.go
index f1751f1..58d2e70 100644
--- a/internal/pwru/tracing.go
+++ b/internal/pwru/tracing.go
@@ -131,17 +131,7 @@ func (t *tracing) trace(coll *ebpf.Collection, spec *ebpf.CollectionSpec,
 ) error {
        // Reusing maps from previous collection is to handle the events together
        // with the kprobes.
-       replacedMaps := map[string]*ebpf.Map{
-               "events":          coll.Maps["events"],
-               "print_stack_map": coll.Maps["print_stack_map"],
-       }
-       if outputSkb {
-               replacedMaps["print_skb_map"] = coll.Maps["print_skb_map"]
-       }
-       if outputShinfo {
-               replacedMaps["print_shinfo_map"] = coll.Maps["print_shinfo_map"]
-       }
-       opts.MapReplacements = replacedMaps
+       opts.MapReplacements = coll.Maps
 
        var errg errgroup.Group
```

They use same bpf maps:

```bash
# bpftool p
3968: kprobe  name kprobe_multi_skb_3  tag 01ea9039709969c3  gpl
        loaded_at 2024-11-27T10:07:34+0800  uid 0
        xlated 7528B  jited 4003B  memlock 12288B  map_ids 24630,24633,24634,24632,24631,24629,24625,24626,24627,24628,24635,24623
        btf_id 27448
3975: tracing  name fentry_xdp  tag 5dae1059a9cdb3e1  gpl
        loaded_at 2024-11-27T10:07:34+0800  uid 0
        xlated 4544B  jited 2524B  memlock 8192B  map_ids 24630,24634,24635,24633
        btf_id 27454
```

Reply:
Still `pwru --filter-track-skb-by-stackid --filter-trace-tc  --output-caller 'src port 19233 and tcp[tcpflags]=tcp-syn'` didn't work as expected. Even more bugs popped up when I tested `opts.MapReplacements = coll.Maps`: tc bpf symbols solution went funny, I saw `__per_cpu_start(tc)` in the output.

I'll go for pinned map solution for now, (+pin path cleanup, +bpf mount). The pin path also requires changes in https://github.com/cilium/pwru?tab=readme-ov-file#running-on-kubernetes, I try not to forget.

Reply:
>  saw __per_cpu_start(tc) in the output.

Because reuse the `.rodata` map. It can be fixed by 

```diff
diff --git a/internal/pwru/tracing.go b/internal/pwru/tracing.go
index f1751f1..0ebdd1c 100644
--- a/internal/pwru/tracing.go
+++ b/internal/pwru/tracing.go
@@ -7,6 +7,7 @@ import (
        "errors"
        "fmt"
        "log"
+       "maps"
        "sync"
 
        "github.com/cilium/ebpf"
@@ -131,16 +132,8 @@ func (t *tracing) trace(coll *ebpf.Collection, spec *ebpf.CollectionSpec,
 ) error {
        // Reusing maps from previous collection is to handle the events together
        // with the kprobes.
-       replacedMaps := map[string]*ebpf.Map{
-               "events":          coll.Maps["events"],
-               "print_stack_map": coll.Maps["print_stack_map"],
-       }
-       if outputSkb {
-               replacedMaps["print_skb_map"] = coll.Maps["print_skb_map"]
-       }
-       if outputShinfo {
-               replacedMaps["print_shinfo_map"] = coll.Maps["print_shinfo_map"]
-       }
+       replacedMaps := maps.Clone(coll.Maps)
+       delete(replacedMaps, ".rodata")
        opts.MapReplacements = replacedMaps
 
        var errg errgroup.Group
```

> Still `pwru --filter-track-skb-by-stackid --filter-trace-tc  --output-caller 'src port 19233 and tcp[tcpflags]=tcp-syn'` didn't work as expected.

Weird. It seems we may have missed something along the way.

Reply:
Thank you :heart: